suite/tests/client-interface/drbbdup-test.dll.c

```diff
@@ -214,6 +214,18 @@ instrument_instr(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
         CHECK(is_first_nonlabel, "should be first non label");
     }
 
+    /* Ensure the drmgr emulation API works for a final special instr where "instr" !=
+     * "where".  The emulation shows up on the label prior to the "where" for the last
+     * instr so we can't compare emul to instr; instead we make sure the emul returned
+     * is never a jump to a label.
+     */
```

Comment:
What about other emulation done by drbbdup clients which include a jump to label? Will confusion arise? I think it would be best to have a drbbdup client also do some from of emulation and then check there are no mismatches

Reply:
I don't understand: this test is using this check for a jump-label as a sanity check for base drbbdup.  A regular client is not going to have a check like this: it is specific to this test.  So if someone added emulation that somehow pointed at a jump-label (ignoring that that doesn't really make sense: why would being-emulated app code have a jump to a label), yes it would thwart this test-specific check: but that is not relevant; that is not part of this test.

The separate topic of ensuring that the drbbdup emulation markers interoperate with other emulation markers: I was going to say that since the drbbdup emulation markers are around its own synthetic instruction it is not possible to overlap with any other emulation, but there is DR_EMULATE_REST_OF_BLOCK -- for that drbbdup should probably not insert any markers at all.  But that's where the drmemtrace errors came from, rep string, which sets that flag: so now it is confusing as to why the rest-of-block flag didn't solve things there.  Plus, this bring up whether the rest-of-block inside drmgr then messes up the subsequent drbbdup cases: a rest-of-block in the first case then applies to *all* the app instructions in all the subsequent cases??  Another drbbdup incompatibility it seems like...

Reply:
> But that's where the drmemtrace errors came from, rep string, which sets that flag: so now it is confusing as to why the rest-of-block flag didn't solve things there.

Scratch that, that's wrong: the emulation errors were coming from every block-final branch, not rep string.  The rep string problem is the crash in PR #5393 which seems to be unrelated to emulation markers but is instead from something to do with drreg control flow.

Reply:
For DR_EMULATE_REST_OF_BLOCK: we can solve that by having drbbdup insert an emulation end label at the end of each case.  drmgr will end a region if it sees that, even if the flag is set.

Reply:
> I don't understand: this test is using this check for a jump-label as a sanity check for base drbbdup. A regular client is not going to have a check like this: it is specific to this test. So if someone added emulation that somehow pointed at a jump-label (ignoring that that doesn't really make sense: why would being-emulated app code have a jump to a label), yes it would thwart this test-specific check: but that is not relevant; that is not part of this test.

Yeah, sorry about that (I had my security hat on); I was trying to get to the interpolation concern and was think of how would I extend this test-case to break those checks.

Reply:
For DR_EMULATE_REST_OF_BLOCK: we can solve that by having drbbdup insert an emulation end label at the end of each case.  drmgr will end a region if it sees that, even if the flag is set.

Assuming that reasoning over the end label is control-flow sensitive, I think that should work.